convert_hf_to_gguf.py

```diff
@@ -373,6 +373,29 @@ def from_model_architecture(cls, arch: str) -> type[Model]:
         except KeyError:
             raise NotImplementedError(f'Architecture {arch!r} not supported!') from None
 
+    def does_token_look_special(self, token: str | bytes) -> bool:
+        if isinstance(token, (bytes, bytearray)):
+            token_text = token.decode(encoding="utf-8")
+        elif isinstance(token, memoryview):
+            token_text = token.tobytes().decode(encoding="utf-8")
+        else:
+            token_text = token
+
+        # Some models mark some added tokens which ought to be control tokens as not special.
+        # (e.g. command-r, command-r-plus, deepseek-coder, gemma{,-2})
+        seems_special = token_text in (
+            "<pad>",  # deepseek-coder
+            "<mask>", "<2mass>", "[@BOS@]",  # gemma{,-2}
+        )
+
+        seems_special = seems_special or (token_text.startswith("<|") and token_text.endswith("|>"))
+        seems_special = seems_special or (token_text.startswith("<｜") and token_text.endswith("｜>"))  # deepseek-coder
+
+        # TODO: should these be marked as UNUSED instead? (maybe not)
+        seems_special = seems_special or (token_text.startswith("<unused") and token_text.endswith(">"))  # gemma{,-2}
```

Comment:
should things like this be defined in the conversion script under the specific model to avoid accidental false hits? if, for some weird reason, a model comes around with a non-special token that starts with <|, would be annoying to avoid that

maybe does_token_look_special should take in 2 lists: 1 list of strings of known special tokens, and a list of tuples of starts/ends with tokens

So for gemma2, we'd call it with:

```
special_tokens = ["<mask>", "<2mass>", "[@BOS@]"]
special_tags = [("<unused", "|>")]

self.does_token_look_special(token, special_tokens, special_tags)
```

and then here we'd have:
```
seems_special = token_text in special_tokens
for start_tag, end_tag in special_tags:
  seems_special = seems_special or (token_text.startswith(start_tag) and token_text.endswith(end_tag))

return seems_special
```

Reply:
oh i realize I misread the structure of the code, hmmm.. still not impossible but would have to be passed at a higher level

Reply:
> if, for some weird reason, a model comes around with a non-special token that starts with <|, would be annoying to avoid that

This only affects added tokens either `added_tokens` in `tokenizer.json` or `added_tokens_decoder` from `tokenizer_config.json`, so it **should not affect normal tokens** starting with `<|` in any way. And added tokens starting with `<|` and ending with `|>` are arguably always control tokens; this was added pretty much *because* some model makers wrongly marked those as non-special (notably, `<|User|>`, `<|Assistant|>` and `<|EOT|>` in `deepseek-coder` are supposedly non-special).

I did not yet notice any conflict in the `added_tokens` to justify making model-specific checks instead of always checking for all known "special-but-arent-marked-special" tokens.